[16.0][MIG] resource_booking: Migration to 16.0

norlinhenrik opened the pull request:
Improvements in this migration:

- Booking: Refactored _get_available_slots() works also when each slot is multiple days
- Booking Type: new field slot_duration
- Contact: button -> Bookings
- Combination -> Bookings -> Create booking: Set the default combination from the navigation
- Booking: may have multiple contacts
- Booking: When type_id is missing (booking from Timeline view), get empty Intervals (instead of error)
- Booking: Search on combination
- Booking: Refactored _availability_is_fitting()
- Booking: List view with hidden contacts (the new field)
- Work intervals (with migration): To span multiple days, set hour_to = 24 instead of 23:59
- XML renaming (with migration)
- #100 

Later PRs for version 15.0 of `resource_booking` (104, 106, 110, 111) are not included in this PR.

Commit:
[ADD] resource_booking: new app to manage bookings
This module adds a new app to allow you to book resource combinations in given
schedules.

Example use cases:

* Management of consultations in a clinic.
* Salesman appointments.
* Classroom and projector reservations.
* Hotel room booking.

Among the things you can do:

* Specify the type of booking, which includes a calendar of availability.
* Specify which resources can be booked together. All of them must be free to be booked.
* Place pending bookings, effectively giving permissions to someone to see the availability calendar and choose one slot.
* Partners can do that from their portals.
* If a partner has no user, he can still do the same via a tokenized URL.
* Backend users can also do that from the backend.
* Booking lifecycle with computed states.
* Automatic meeting creation and deletion.
* Automatic conflict detection.
* Deadline to block modifications.

@Tecnativa TT28201

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (190 of 190 strings)

Translation: calendar-12.0/calendar-12.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-12-0/calendar-12-0-resource_booking/es/

Commit:
[IMP] resource_booking: filter/group by date

TT29508

Commit:
[FIX] resource_calendar: ignore past or unconfirmed bookings when updating calendars

Without this patch, users couldn't change a calendar schedule if there were past or unconfirmed bookings that wouldn't fit in it.

Excluding those bookings from the check fixes the situation.

We also check that, to confirm a booking, it must fit in the calendar (because now it can happen that, in the time that has passed since the booking was scheduled until it is confirmed, the calendar changes).

@Tecnativa TT29509

Commit:
[FIX] resource_booking: always notify in resource TZ

The notifications emitted to the resource booking requester must always be in the same TZ as the resource booking itself.

For example, if you book one hotel room in the other side of the world, a notification in your own TZ is confusing.

Besides, res.partner created from website_sale are created with `tz=False`, making it even more confusing.

@Tecnativa TT30331

joao-p-marques committed:
[FIX] resource_booking: Limit constraint only to future bookings

The constraint that checks the schedule of a resource booking is currently being applied to all the bookings, including past ones.
As the resource combination or associated calendars might change regularly and trigger a recomputation of this, such change might take a very long time.
Plus, the calendar restrictions might change, trigger a recompute of the constraint and detect bookings that can't be assigned, which makes no sense when they already happened.

This applies it only to future bookings, ignoring past ones.

TT30478

Commit:
[IMP] resource_booking: black, isort, prettier

Commit:
[MIG] resource_booking: migrate to v13

- Standard v13 changes.
- Some onchanges removed in favor of computes.

@Tecnativa TT30987

Commit:
[IMP] resource_booking: simplified time format

It's very unlikely that you need to book resources in a seconds-based precision.

This simplifies portal UI. Still, declared as a variable in case you need some customizations downstream.

@Tecnativa TT31063

Commit:
[IMP] resource_booking: configurable auto-assigned combination and location

Thanks to this patch, you will be able to more predictably block the chosen resource combination, with the added checkbox.

Also, the location field is propagated to `resource.booking` records, and synced from there to `calendar.event` if needed. This way, you can alter the location before sending the portal invitation.

@Tecnativa TT30987

Commit:
[IMP] resource_booking: use duration field instead of stop

From now on, `resource.booking.type` duration represents the default duration that will be applied to new `resource.booking` of that type, instead of the hardcoded duration for every new booking.

This means that you can create a new booking in pending state and define a custom duration before sending the portal link to customer.

Thus, the available schedule slots in portal will be based on the type duration, whereas only those that have enough time left to complete the booking duration will be displayed.

@Tecnativa TT30987

Commit:
[FIX] resource_booking: "Involves me" filter

The filter wasn't working fine due to the `auto_join=True` set in `meeting_id`. It was never displaying pending bookings.

Besides, it wasn't necessary to keep it as a field because it was displayed nowhere and the search could be done directly in the view. This way, there's less python code to maintain and we disable the possibility of having a negative domain, which enters doomed scenarios when x2many fields are involved (see https://github.com/odoo/odoo/issues/43957).

@Tecnativa TT30987

Commit:
[IMP] resource_booking: compatible with hr_holidays_public

Adding `.with_context(exclude_public_holidays=True)` in needed places to let `hr_holidays_public` do its magic and never allow allocating a booking during a public holidays when that module is properly installed and configured.

@Tecnativa TT30987

Commit:
[IMP] resource_booking: optional name propagated to meeting

- Allow users to optionally define a name for the resource booking.
- Display that name in portal if set.
- Sync it with meeting name when creating it.

@Tecnativa TT30987

Commit:
[FIX] resource_booking: error when recurrent meetings exist

Without this patch, when entering the Resource Bookings menu, if the current user attended any recurrent meeting, it failed with this error:

```
Server application error
 Error code: 200
 Error message: Odoo Server Error
 Error data message:
 invalid input syntax for integer: "82-20210726143000"
LINE 1: ... AND  ("resource_booking"."meeting_id" in (3,5,81,'82-202107...
                                                             ^

 Error data debug:
 Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/auto/addons/web/controllers/main.py", line 1339, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/auto/addons/web/controllers/main.py", line 1331, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 383, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 356, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 4936, in search_read
    records = self.search(domain or [], offset=offset, limit=limit, order=order)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 1648, in search
    res = self._search(args, offset=offset, limit=limit, order=order, count=count)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 4499, in _search
    self._cr.execute(query_str, where_clause_params)
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 173, in wrapper
    return f(self, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 250, in execute
    res = self._obj.execute(query, params)
psycopg2.DataError: invalid input syntax for integer: "82-20210726143000"
LINE 1: ... AND  ("resource_booking"."meeting_id" in (3,5,81,'82-202107...
                                                             ^
```

@Tecnativa TT30987

pedrobaeza committed:
[FIX] resource_booking: Proper calendar.event unlink signature

Upstream ``unlink`` method in calendar.event adds this keyword argument.

Commit:
[FIX] resource_booking: access error on normal calendar

When a user had no resource_booking permissions and opened a calendar event linked to a resource booking, he got an unnecessary access error.

@Tecnativa TT31238

Commit:
[IMP] resource_booking: autoconfirm meeting attendees

- If RBC (resource.booking.combination) was auto-assigned, attendees generated from RBC human resources are expected to confirm attendance manually.
- If RBC was handpicked, those attendees are auto-confirmed.
- All those confirmations are done before sending invitations, so invitations contain the correct attendance status.
- If requester books from portal view, he's also confirmed before sending invitation, for the same reason.

@Tecnativa TT31239

Commit:
[FIX] resource_booking: allow combination with other kind of leaves

When `_leave_intervals_batch()` in `resource.calendar` is extended by other modules that add more leaves for any reason, the returned recordset must be of the `resource.calendar.leaves` model. Otherwise, Odoo fails with:

    TypeError: '<' not supported between instances of 'resource.resource' and 'resource.calendar.leaves'

The resource didn't have a meaning here, so changing it to avoid bug.

@Tecnativa TT31249

Commit:
[FIX] resource_booking: display location and duration on portal

Missing part from https://github.com/OCA/calendar/pull/34:

- Display booking duration and location on portal, even on pending state.
- Improve portal test to contemplate new behaviors.

@Tecnativa TT31250

Commit:
[IMP] resource_booking: improve auto-assignment readability

@Tecnativa TT31239

Commit:
[IMP] resource_booking: add organizer

This allows better filtering while in the normal calendar view.

By default, the organizer will be whoever created the booking. Calendar invitations will go out in his name.

Also, avoid sending calendar event modification notifications if the modification isn't the schedule.

@Tecnativa TT31240

resource_booking 13.0.2.3.0

Commit:
[FIX] resource_booking: suggest involved partners as followers

Includes those coming from requester and from resource combination.

@Tecnativa TT31901

abdawone committed:
Added translation using Weblate (French (France))

Currently translated at 97.0% (196 of 202 strings)

Translation: calendar-13.0/calendar-13.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-13-0/calendar-13-0-resource_booking/fr_FR/

abdawone committed:
Added translation using Weblate (French)

Currently translated at 1.4% (3 of 202 strings)

Translation: calendar-13.0/calendar-13.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-13-0/calendar-13-0-resource_booking/fr/

Commit:
[FIX] resource_booking: autosubscribe resource partners

When some partner is a resource assigned to a booking, he should get auto-subscribed.

Inspired in https://github.com/OCA/calendar/pull/42, but I'm changing that behavior.

@Tecnativa TT32148

Commit:
[FIX] resource_booking: correct assignment message

By using the upstream `mail.message_user_assigned` template, we were getting this exception:

```
odoo.addons.base.models.qweb.QWebException: 'resource.booking' object has no attribute 'user_id'
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/qweb.py", line 348, in _compiled_fn
    return compiled(self, append, new, options, log)
  File "<template>", line 1, in template_2601_10519
AttributeError: 'resource.booking' object has no attribute 'user_id'
```

So I provide a specific template for this kind of subscription notification, which doesn't use that `user_id` field. Besides, it explains better why you're getting it.

@Tecnativa TT32190

Commit:
[FIX] resource_booking: suggest requester to messaging

Forward-port of https://github.com/OCA/calendar/pull/51.

Includes a modification for the migration script of 13.0.1.0.0 (if you migrate from v12, you'll need it; if you migrate from v13, it will be automatic).

@Tecnativa TT31901

Commit:
[FIX] resource_booking: allow creating if type has categories

Without this patch, if you create a resource booking for a type that has categories (tags), it will fail with `CacheMiss` error.

@Tecnativa TT32552

Commit:
[FIX] resource_booking: free meetings by default

If I'm creating bookings for my colleagues, I'll be the owner of all of the meetings. But that doesn't mean I'll be busy.

With this patch, only when I'm a booked resource will the event be set as busy.

@Tecnativa TT32666

jguenat committed:
[FIX]start_next_previous

Commit:
[IMP] resource_booking: faster tests

Apply lessons from https://youtu.be/lT___hG110Q?t=23285 and disable unnecessary mail stuff.

Before:

    odoo.addons.resource_booking.tests.test_backend tested in 16.18s, 9124 queries
    odoo.addons.resource_booking.tests.test_portal tested in 16.42s, 6685 queries

After:

    odoo.addons.resource_booking.tests.test_backend tested in 13.68s, 7873 queries
    odoo.addons.resource_booking.tests.test_portal tested in 15.17s, 5972 queries

@Tecnativa TT31324

Commit:
[FIX] Avoid sending notifications when public user

Public users usually don't have an email, so they hit [this error][1].

No test added because anyways the tests in `website_sale_resource_booking` already fail without this patch, and it's hard to find a use case without it.

@Tecnativa TT31328

[1]: https://github.com/odoo/odoo/blob/e7dd314e53a4260dd1494cbcb7dcef1c407e2a8a/addons/mail/models/mail_message.py#L36

chienandalu committed:
[FIX] resource_booking: performance

- Calendar view performance: We need a stop parameter. Otherwise, the calendar view
  could be trying to render tons of records making the client's browser have a hard time.

- Performance with recurrent events:
  `rrule` isn't very performant on iterations. This is used on calendar
  events to compute the virtual events set.

  We want to avoid calling it too much.

  Real case with website_sale_resource_booking (thousands of virtual events):

  before: /shop/booking/<id>/schedule: ~25 seconds
  after: /shop/booking/<id>/schedule: ~6 seconds

  before: /shop/booking/<id>/confirm: ~20 seconds
  after: /shop/booking/<id>/confirm: ~3 seconds

  before: /shop/checkout: ~18 seconds
  after: /shop/checkout: ~2 seconds

  There's room for improvement anyway

- Restrict the events which our resource is participating in to avoid
  loading recurrent events as much as possible. This is really more
  performant in v13. A simple showcase:

  >>> timeit(
          lambda: env["calendar.event"].with_context(active_test=True).search(
              [("partner_ids", "=", partner.id)]
          ), number=1
      )
  0.06541907804785296
  >>> timeit(
          lambda: env["calendar.event"].with_context(active_test=True).search([])
          , number=1
      )
  1.194712486001663

pedrobaeza committed:
[IMP] resource_booking: pre-commit stuff

pedrobaeza committed:
[MIG] resource_booking: Migration to 14.0

- Standard procedure
- No more virtual ids
- Adapt tests
- FIX: Proper show_as meeting generation
- IMP: Add name to tree view

TT32344

[UPD] Update resource_booking.pot

[UPD] README.rst

carmenbianca committed:
[IMP] resource_booking: Allow a booking to span more than one calendar day

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] resource_booking: Create resource.is_available() convenience function

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk committed:
[IMP] add description field to bookings

victoralmau committed:
[IMP] resource_booking: Change the position of the description field for a correct display.

resource_booking 14.0.1.2.0

pedrobaeza committed:
[FIX] resource_booking: Don't fail with multiple internal attendees

Steps to reproduce:

- Create a resource combination with 2 user resources.
- Create a type with such resource combination.
- Create a resouce booking forcing manually such resource combination.
- Book from the portal a slot for it.

Result: Singleton error

File "/opt/odoo/auto/addons/resource_booking/controllers/portal.py", line 134, in portal_booking_confirm
  booking_form.start = when_naive
File "/opt/odoo/custom/src/odoo/odoo/tests/common.py", line 1979, in __exit__
  self.save()
File "/opt/odoo/custom/src/odoo/odoo/tests/common.py", line 1996, in save
  r.write(values)
File "/opt/odoo/auto/addons/resource_booking/models/resource_booking.py", line 509, in write
  self._sync_meeting()
File "/opt/odoo/auto/addons/resource_booking/models/resource_booking.py", line 337, in _sync_meeting
  _self.env["calendar.event"].create(to_create)
File "<decorator-gen-216>", line 2, in create
File "/opt/odoo/custom/src/odoo/odoo/api.py", line 348, in _model_create_multi
  return create(self, arg)
File "/opt/odoo/auto/addons/crm/models/calendar.py", line 40, in create
  events = super(CalendarEvent, self).create(vals)
File "<decorator-gen-201>", line 2, in create
File "/opt/odoo/custom/src/odoo/odoo/api.py", line 348, in _model_create_multi
  return create(self, arg)
File "/opt/odoo/auto/addons/resource_booking/models/calendar_event.py", line 65, in create
  records += super(
File "<decorator-gen-183>", line 2, in create
File "/opt/odoo/custom/src/odoo/odoo/api.py", line 347, in _model_create_multi
  return create(self, [arg])
File "<decorator-gen-148>", line 2, in create
File "/opt/odoo/custom/src/odoo/odoo/api.py", line 348, in _model_create_multi
  return create(self, arg)
File "/opt/odoo/auto/addons/calendar/models/calendar_event.py", line 763, in create
  vals_list = [
File "/opt/odoo/auto/addons/calendar/models/calendar_event.py", line 764, in <listcomp>
  dict(vals, attendee_ids=self._attendees_values(vals['partner_ids']))
File "/opt/odoo/auto/addons/resource_booking/models/calendar_event.py", line 102, in _attendees_values
  ctx_partner_id = rb.combination_id.resource_ids.user_id.partner_id.id
File "/opt/odoo/custom/src/odoo/odoo/fields.py", line 3822, in __get__
  raise ValueError("Expected singleton: %s" % record)
ValueError: Expected singleton: res.partner(6, 4771)

This is because there can be several linked users, but the code is
assuming that there's only one.

This commits fixes such possibility.

resource_booking 14.0.1.2.1

eugenichafer committed:
Added translation using Weblate (Catalan)

eugenichafer committed:
Translated using Weblate (Catalan)

Currently translated at 6.6% (14 of 209 strings)

Translation: calendar-14.0/calendar-14.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-14-0/calendar-14-0-resource_booking/ca/

victoralmau committed:
[MIG] resource_booking: Migration to 15.0

TT40302

pedrobaeza committed:
[FIX] resource_booking: Allow non RB users to handle calendar events

Steps to reproduce:

- Create a calendar event
- Edit it
- Add an attendee
- Save it

Result: you get a permission error to resource.booking

This is because there are some operations done on resource bookings
from the calendar events. The solution is to do such operations with
sudo.

resource_booking 14.0.1.2.2

victoralmau committed:
[FIX] resource_booking: Force to send calendar event author notification

carmenbianca committed:
[IMP] resource_booking: Add description to calendar view card

This allows the user to see the summary of an event from the calendar
view.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update resource_booking.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: calendar-15.0/calendar-15.0-resource_booking
Translate-URL: https://translation.odoo-community.org/projects/calendar-15-0/calendar-15-0-resource_booking/

victoralmau committed:
[IMP] resource_booking: Use _prepare_home_portal_values() function.

victoralmau committed:
[FIX] resource_booking: Change _default_company() to self.env.company from booking types

victoralmau committed:
[IMP] resource_booking: Change t-esc directives to t-out according to v15 guidelines

victoralmau committed:
[IMP] resource_booking: Partial change from tour test to js

OCA-git-bot committed:
resource_booking 15.0.1.1.0

norlinhenrik committed:
[MIG] resource_booking: Migration to 16.0

OCA-git-bot committed:
[UPD] README.rst

norlinhenrik commented:
![image](https://github.com/OCA/calendar/assets/11876900/7f144c21-bc74-48ff-9888-562bb5a1f90f)
The partner calendar dropdown menu does not work in version 16.0.
I am not a frontend developer. Is there anyone else who could troubleshoot this?

norlinhenrik commented:
I had to rename `data-toggle` to `data-bs-toggle`. I did the same with some other data attributes, in a separate commit.

pedrobaeza commented:
Check https://github.com/OCA/openupgradelib/pull/338

norlinhenrik commented:
Thank you for the link, @pedrobaeza
How do I use the BS4 to BS5 tool to update a module?

data-target should change to data-bs-target.
Should `t-attf-data-target` change to `t-attf-data-bs-target` ?

pedrobaeza commented:
> How do I use the BS4 to BS5 tool to update a module?

It's more like a mapping reference than a direct tool

> Should t-attf-data-target change to t-attf-data-bs-target ?

Yes, as the prefix `t-attf-` means that the rest of the attribute is parametric.

OCA-git-bot committed:
oca-github-bot post-merge updates

norlinhenrik committed:
[IMP] resource_booking: pre-commit stuff

norlinhenrik commented:
I have included [Allow a booking to span more than one calendar day #100](https://github.com/OCA/calendar/pull/100)

@yajo @pedrobaeza Would you like to review?

norlinhenrik committed:
[FIX] resource_booking: _get_available_slots

[IMP] resource_booking: new field slot_duration

[IMP] resource_booking: Button (partner -> booking)

[IMP] resource_booking: combination -> bookings -> create: default combination

[IMP] resource_booking: A booking may have multiple contacts, e.g. people in a room

[FIX] resource_booking: _get_intervals() when type_id is missing

[IMP] resource_booking: booking: search on combination

[FIX] resource_booking: _availability_is_fitting()

[IMP] resource_booking: booking list view with hidden partner_ids

[FIX] resource_booking: migration: attendance hour_to 23:59 -> 24:00

[IMP] resource_booking: rename_xmlids

[FIX] resource_booking: pre-commit

[FIX] resource_booking: calendar_slot_duration format

[FIX] resource_booking: _get_calendar_context() with correct start / timezone

victoralmau commented:
Please, cherry-pick https://github.com/OCA/calendar/pull/104 (resolves the comment in https://github.com/OCA/calendar/pull/101#issuecomment-1716288482)

stefan-tecnativa commented on resource_booking/models/resource_booking_type.py:
```diff
@@ -0,0 +1,149 @@
+                default_type_id=self.id,
+                # Context used by web_calendar_slot_duration module
+                calendar_slot_duration=FloatTimeParser.value_to_html(
+                    self.duration, False
```
```suggestion
                    self.slot_duration, False
```
You added a slot_duration field intentionally to separe the duration of the booking and the slot in calendar, so this change is consequent. But anyway this change is uselles because without the dependency of `web_calendar_slot_duration` this doesn't work. 
If you look at my PR: https://github.com/OCA/web/pull/2646 , you must notice that the slot_duration that expects is in this format: (00:00:00), but this FloatTimeParser function returns it in this one: (00:00). So you will have to make a conversion from the Float to this first format.

appstogrow replied:
Is (00:00:00) new for v16?
If yes, can you help me with this conversion?
If no, can you show me where the conversion is done for self.duration?

stefan-tecnativa reviewed (changes requested):
Please, solve the `web_calendar_slot_duration` dependency and attend the comments.

@Tecnativa TT45649

stefan-tecnativa commented on resource_booking/__manifest__.py:
```diff
@@ -25,7 +25,6 @@
         "mail",
         "portal",
         "resource",
-        "web_calendar_slot_duration",
```
You can't remove this dependency, it is necessary to maintain the functionallity of the module. Here I post a PR that i made migrating correctly the module: https://github.com/OCA/web/pull/2646

carolinafernandez-tecnativa commented:
Hi how are you? There is one commit missing from 15.0
https://github.com/OCA/calendar/commit/a4548a1b597228ebdc51e8506a3a1a40ab9516ce

flotho reviewed (commented):
Thanks for this migration.
Functional tests in progress
ACL issue for admin :
![image](https://github.com/OCA/calendar/assets/1043935/96af5911-d982-4a6d-a853-f4c14c3a3575)

IMHO, admin user should be able to setup thoses exceptions

flotho commented:
Question regarding the process, 
Does this module could also replace something like calendly ?
In this case, how to share available slot to be booked on the portal and/or on public website ?
Regards

pedrobaeza commented:
Yes, it can act like Calendly. Just click on the Share button of the resource booking.

norlinhenrik commented:
Thank you for all the feedback. I will work on this one of the next days.

norlinhenrik commented:
> Thanks for this migration. Functional tests in progress ACL issue for admin : ![image](https://user-images.githubusercontent.com/1043935/279205903-96af5911-d982-4a6d-a853-f4c14c3a3575.png)
> 
> IMHO, admin user should be able to setup thoses exceptions

Security for Resource Time Off is defined in `resource`. Admin may create global leaves, but not leaves for specific human resources.

norlinhenrik commented:
Should I squash together all my commits, or just those who are highly related?

Improvements in this migration:

- Booking: Refactored _get_available_slots() works also when each slot is multiple days
- Booking Type: new field slot_duration
- Contact: button -> Bookings
- Combination -> Bookings -> Create booking: Set the default combination from the navigation
- Booking: may have multiple contacts
- Booking: When type_id is missing (booking from Timeline view), get empty Intervals (instead of error)
- Booking: Search on combination
- Booking: Refactored _availability_is_fitting()
- Booking: List view with hidden contacts (the new field)
- Work intervals (with migration): To span multiple days, set hour_to = 24 instead of 23:59
- XML renaming (with migration)

JavierIniesta commented:
Hi @norlinhenrik 

Thank you for your efforts.

Could you push again to see if the problem with the pre-commit is resolved? It looks like there was some problem during the process with the nodejs environment.

norlinhenrik commented:
pre-commit is fixed.

Who can suggest a fix for the `slot_duration` format?

unique4 commented:
> pre-commit is fixed.
> 
> Who can suggest a fix for the `slot_duration` format?

If slot_duration purpose is to snap to intervals, then I suggest that default duration should be multiples of slot_duration. Maybe a constraint function here?

unique4 commented on resource_booking/models/resource_booking_type.py:
```diff
@@ -0,0 +1,149 @@
+        return combinations
+
+    def action_open_bookings(self):
+        FloatTimeParser = self.env["ir.qweb.field.float_time"]
+        return {
+            "context": dict(
+                self.env.context,
+                default_alarm_ids=[(6, 0, self.alarm_ids.ids)],
+                default_description=self.requester_advice,
+                default_duration=self.duration,
+                default_type_id=self.id,
+                # Context used by web_calendar_slot_duration module
+                calendar_slot_duration=FloatTimeParser.value_to_html(
+                    self.slot_duration, False
+                ),
```
@norlinhenrik can use this base parser
```suggestion
        DurationParser = self.env["ir.qweb.field.duration"]
        return {
            "context": dict(
                self.env.context,
                default_alarm_ids=[(6, 0, self.alarm_ids.ids)],
                default_description=self.requester_advice,
                default_duration=self.duration,
                default_type_id=self.id,
                # Context used by web_calendar_slot_duration module
                calendar_slot_duration=DurationParser.value_to_html(self.slot_duration, {
                    "unit": "hour",
                    "digital": True,
                }),
```

norlinhenrik replied:
Done

unique4 commented:
@norlinhenrik BTW, if slot_duration purpose is to snap to intervals, then I suggest that default duration should be multiples of slot_duration. Maybe a constraint function here?

norlinhenrik commented:
Please approve or suggest specific code improvements.

#100 is included in this PR. Later PRs for `resource_booking` (104, 106, 110, 111) are not included in this PR.

pedrobaeza commented:
#113 is also interesting here.

By the way, it seems this is still red

rrebollo commented:
Hello, I'm deploying this PR on a local dev environment based on Doodba, looking forward wanting to contribute in any way.
I need to ask: shouldn´t be freezegun added to \_\_manifest\_\_.py as an external dependency. I see it in README.rst and code tests @norlinhenrik. I think I can do something about the red tests too. Do I just fork this branch @norlinhenrik???

norlinhenrik commented:
@rrebollo Feel free to fork my branch and fix the errors. Then give me the link to your repo so I can pull the changes and add them to this PR. Thank you!

rrebollo committed:
[IMP] resource_booking: add freezegun as a python external dependency

[IMP] resource_booking: fix failling BackendCase. test_booking_from_calendar_view

[IMP] resource_booking: fix typo in help parameter for combination_assignment field from resource_booking_type

[IMP] resource_booking: fix failing PortalCase.test_portal_no_bookings

Test was failling cause there was no bookings link on portal home for external users, then no trigger.
Also I had to tweak eslintrc config cause precommit was failling due to ECMA version.

[IMP] resource_booking: No create/unlink call if nothing to do

[FIX] resource_booking: error CalendarQuickCreate title is not a string when scheduling a booking

[FIX] resource_booking: portal responsive booking calendar table going wider and no x scroll

rrebollo commented:
@pedrobaeza [CONTRIBUTING](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst) mentions oca_dependencies.txt which seems apropiate in this case too because web_calendar_slot_duration dependency. Is it an outdated convention???

rrebollo commented:
@norlinhenrik I'm working in
`ERROR odoo odoo.addons.resource_booking.tests.test_backend: FAIL: BackendCase.test_booking_from_calendar_view`

rrebollo commented:
@norlinhenrik my fork is https://github.com/rrebollo/calendar.git branch 16.0. Added freezegun python dependency, fixed failling BackendCase.test_booking_from_calendar_view. I will verify the remaining failling test. I was thinking to look after the fields.Command usage. What do U think???

rrebollo commented:
@norlinhenrik PortalCase.test_portal_no_bookings solved too on my fork. Fixes some typo and I had to deal with eslintrc precommit hook failling 'cause ECMA version. Hope U get me some feedback soon! Have a nice weekend. @pedrobaeza I also incorported your code from #113

ayushin commented:
what does it take to get this over the finish line?

norlinhenrik commented:
Thank you @rrebollo I have pushed your commits to the PR now.

norlinhenrik commented:
> @pedrobaeza [CONTRIBUTING](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst) mentions oca_dependencies.txt which seems apropiate in this case too because web_calendar_slot_duration dependency. Is it an outdated convention???

I would like to update the documentation about `oca_dependencies.txt.` Please join the discussion [here](https://github.com/OCA/odoo-community.org/pull/148).

rrebollo commented:
@norlinhenrik I was doing some testing using runboat build on the last commit to this PR. Something bizarre it's happening using a tokenized url to portal to schedule the book: It gets stuck on December 2023. See what I mean. First I'm showing the settings (Basically everyone is in the same TZ), later the "problem". I'm using a negative timezone (-5 I think):

![Peek Portal Schedule booking stuck on same month](https://github.com/OCA/calendar/assets/38138332/6db1b8df-3424-420d-a199-062ce4f7483c)

Surely the core of the issue is:
[here](https://github.com/ows-cloud/calendar/blob/813c223fdc2cf3ce2a54d7e656077fc1c7cb4afb/resource_booking/models/resource_booking.py#L483-L513)


```python
    def _get_calendar_context(self, year=None, month=None, now=None):
        """Get the required context for the calendar view in the portal.


        See the `resource_booking.scheduling_calendar` view.


        :param int year: Year of the calendar to be displayed.
        :param int month: Month of the calendar to be displayed.
        :param datetime now: Represents the current datetime.
        """
        month1 = relativedelta(months=1)
        now = now or fields.Datetime.now()
        year = year or now.year
        month = month or now.month
        start = datetime(year, month, 1)
        start, now = (
            fields.Datetime.context_timestamp(self, dt) for dt in (start, now)
        )
        start = start.replace(hour=0, minute=0, second=0, microsecond=0)
        lang = self.env["res.lang"]._lang_get(self.env.lang or self.env.user.lang)
        weekday_names = dict(lang.fields_get(["week_start"])["week_start"]["selection"])
        booking_duration = timedelta(hours=self.duration)
        slots = self._get_available_slots(start, start + month1 + booking_duration)
        return {
            "booking": self,
            "calendar": calendar.Calendar(int(lang.week_start) - 1),
            "now": now,
            "res_lang": lang,
            "slots": slots,
            "start": start,
            "weekday_names": weekday_names,
        }
``` 

It isn't clear to me why:
```python
start, now = (
            fields.Datetime.context_timestamp(self, dt) for dt in (start, now)
        )
``` 
should'nt the target here be January 1st, 2024 no matter TZ? @pedrobaeza Am I getting things wrong?
The "issue" happens to be "solved" with
```
❯ git diff
diff --git a/resource_booking/models/resource_booking.py b/resource_booking/models/resource_booking.py
index effe7d9..ccea10d 100644
--- a/resource_booking/models/resource_booking.py
+++ b/resource_booking/models/resource_booking.py
@@ -508,7 +508,7 @@ class ResourceBooking(models.Model):
             "now": now,
             "res_lang": lang,
             "slots": slots,
-            "start": start,
+            "start": fields.Date.from_string(f'{year}-{month}-01'),
             "weekday_names": weekday_names,
         }
 
diff --git a/resource_booking/templates/portal.xml b/resource_booking/templates/portal.xml
index 8931d7e..54b82fd 100644
--- a/resource_booking/templates/portal.xml
+++ b/resource_booking/templates/portal.xml
@@ -44,7 +44,7 @@
                     <tr>
                         <th class="text-left">
                             <a
-                                t-if="start > now"
+                                t-if="start > now.date()"
                                 t-att-href="booking.get_portal_url(suffix='/schedule/%d/%d' % (start_previous.year, start_previous.month))"
                                 class="btn btn-secondary"
                                 title="Previous month"
```

norlinhenrik commented:
`start` should be the very beginning of the month, in the timezone of self.env["tz"].
This is fixed in the latest commit.

rrebollo commented:
@norlinhenrik partner auto-subcription to bookings looks like isn't working either [here](https://github.com/OCA/calendar/commit/e9437a6827f90c7d95b0d0ff917e4538bb379630). I'm taking this.
Edited:  auto subscription doesn't work on booking creation. It's fine at update.

rrebollo commented:
> @norlinhenrik partner auto-subcription to bookings looks like isn't working either [here](https://github.com/OCA/calendar/commit/e9437a6827f90c7d95b0d0ff917e4538bb379630). I'm taking this. Edited: auto subscription doesn't work on booking creation. It's fine at update.

@norlinhenrik I think it is intended because in related calendar events followers are rigthtly settled. So I think no work is required here.

rrebollo commented:
@norlinhenrik I made an small fix. When trying to schedule a booking from booking form view but no name has been given to the booking it raises an OWl error when clicking on the calendar view. I created a PR to your branch. Please check it out.

ayushin commented:
not to rush anybody, but lets keep the momentum!

i'm willing to help in any way i can to get this merged asap

norlinhenrik commented:
@ayushin If you have tested and think it is ready to merge, then click on Files changed -> Review changes -> Approve.

jelenapoblet reviewed (commented):
It's working just fine, i've been testing it out on runboat. The only thing that bugs me a little it's on the smartphone view.
![WhatsApp Image 2024-02-05 at 10 32 45 PM](https://github.com/OCA/calendar/assets/119665889/7ea023d2-3408-4870-9a7b-54ad9f1aced0)

rrebollo commented:
> It's working just fine, i've been testing it out on runboat. The only thing that bugs me a little it's on the smartphone view. ![WhatsApp Image 2024-02-05 at 10 32 45 PM](https://private-user-images.githubusercontent.com/119665889/302447198-7ea023d2-3408-4870-9a7b-54ad9f1aced0.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMTEzMDIsIm5iZiI6MTcwNzIxMTAwMiwicGF0aCI6Ii8xMTk2NjU4ODkvMzAyNDQ3MTk4LTdlYTAyM2QyLTM0MDgtNDg3MC05YTdiLTU0YWQ5ZjFhY2VkMC5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMDZUMDkxNjQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWFmY2JjNjFhMzYxNWI4OTkyZGM1YzY0MDI4YTg5YjRlZTZiN2UyNjViNDhjMDI4YzdiNzBlZjY2OWNhNTIwYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Km_B-qTzshlmwJ9PvLBXniq2EZvAxyoNzFKkXH6_-GI)

It's happening here too. Any known frontend dev to assit us???

rrebollo commented:
@jelenapoblet I think I "fixed" the calendar going wider. According to Bootstrap 5.1 doc .table-responsive-md should be in a div as parent for the table element with .table. @norlinhenrik PR https://github.com/ows-cloud/calendar/pull/1.

pedrobaeza commented:
/ocabot migration resource_booking

Please finish your reviews.

norlinhenrik commented:
@rrebollo I have included your latest commit.

pedrobaeza commented:
Great, please squash all the commits by author into their migration one for being mergeable.

norlinhenrik commented:
Done

pedrobaeza commented on resource_booking/__manifest__.py:
```diff
@@ -0,0 +1,53 @@
+        "python": [
+            # Used implicitly
+            "cssselect",
+            "freezegun",
```
This is not correct. This dependency is not for the module, but for tests. You must remove it from here and put it in `test_requirements.txt` file in root.

rrebollo replied:
@norlinhenrik I have pushed to https://github.com/ows-cloud/calendar/pull/1 a commit fixing this

pedrobaeza replied:
`cssselect` may be needed for the module itself. I was talking about `freezegun`. Anyway, the CI is still red.

norlinhenrik commented:
All checks have passed now.

pedrobaeza commented on resource_booking/models/resource_booking.py:
```diff
@@ -0,0 +1,753 @@
+    )
+    name = fields.Char(index=True, help="Leave empty to autogenerate a booking name.")
+    description = fields.Html()
+    partner_id = fields.Many2one(
```
This is not correct according v15:

https://github.com/OCA/calendar/blob/15.0/resource_booking/models/resource_booking.py#L126

The compute/inverse is in this field, not in `partner_ids`.

norlinhenrik replied:
`partner_ids` was added to 15.0 after I suggested it for 16.0. Probably I don't need this field after all.

@pedrobaeza Feel free to propose a new commit to do the correction. You may remove `partner_ids` if you want.

norlinhenrik replied:
But I think the field may be practcal in some scenarios, even though I don't need it right now.

pedrobaeza replied:
@victoralmau please resume this migration including all the 15.0 stuff. Please complete and include #114 as well.

victoralmau commented:
Superseed by https://github.com/OCA/calendar/pull/120

pedrobaeza commented:
As the other has preserved attribution, closing in favor of it.
